### PR TITLE
[AMDGPU] Fix an issue that wrong index is used in calculation of byte provider when the op is extract_vector_elt

### PR DESCRIPTION
**Pull request by shiltian**
Fixes: SWDEV-460097

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 1]
```diff
@@ -12071,12 +12071,8 @@ calculateByteProvider(const SDValue &Op, unsigned Index, unsigned Depth,
...
     auto VecIdx = IdxOp->getZExtValue();
     auto ScalarSize = Op.getScalarValueSizeInBits();
-    if (ScalarSize != 32) {
-      Index = ScalarSize == 8 ? VecIdx : VecIdx * 2 + Index;
-    }
-
-    return calculateSrcByte(ScalarSize == 32 ? Op : Op.getOperand(0),
-                            StartingIndex, Index);
+    return calculateSrcByte(Op.getOperand(0), StartingIndex,
+                            ScalarSize / 8 * VecIdx + Index);
```
Comment (shiltian) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:12075]: I feel I might miss some corner case here but I can't figure that out. However, this approach does look pretty straightforward.

**Comment by jrbyrnes** [comment]
I see a regression of class:

define hidden void @new(ptr addrspace(1) %in0, ptr addrspace(1) %in1, ptr addrspace(1) %out0) {
  %vec0 = load <4 x i32>, ptr addrspace(1) %in0, align 4
  %vec1 = load <4 x i32>, ptr addrspace(1) %in1, align 4
  %v0e1 = extractelement <4 x i32> %vec0, i64 1
  %v1e3 = extractelement <4 x i32> %vec1, i64 3
  %bc0 = bitcast i32 %v0e1 to <4 x i8>
  %bc1 = bitcast i32 %v1e3 to <4 x i8>
  %shuffle0_0 = shufflevector <4 x i8> %bc0, <4 x i8> %bc1, <4 x i32> <i32 3, i32 7, i32 4, i32 4>
  store <4 x i8> %shuffle0_0, ptr addrspace(1) %out0, align 4
  ret void
}

File: llvm/test/CodeGen/AMDGPU/match-perm-extract-vector-elt-bug.ll [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+; GFX11-NEXT:    s_nop 0
+; GFX11-NEXT:    s_sendmsg sendmsg(MSG_DEALLOC_VGPRS)
+; GFX11-NEXT:    s_endpgm
+entry:
+  %test.kernarg.segment = call nonnull align 16 dereferenceable(280) ptr addrspace(4) @llvm.amdgcn.kernarg.segment.ptr()
+  %srcA.kernarg.offset4 = bitcast ptr addrspace(4) %test.kernarg.segment to ptr addrspace(4)
+  %srcA.load = load ptr addrspace(1), ptr addrspace(4) %srcA.kernarg.offset4, align 16
+  %dst.kernarg.offset = getelementptr inbounds i8, ptr addrspace(4) %test.kernarg.segment, i64 16
+  %dst.load = load ptr addrspace(1), ptr addrspace(4) %dst.kernarg.offset, align 16
+  %0 = tail call ptr addrspace(4) @llvm.amdgcn.implicitarg.ptr()
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/match-perm-extract-vector-elt-bug.ll:106]: Use named values in tests. You can also probably reduce most of this away

File: llvm/test/CodeGen/AMDGPU/match-perm-extract-vector-elt-bug.ll [hunk 3]
```diff
@@ -0,0 +1,126 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -verify-machineinstrs -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 < %s | FileCheck -check-prefix=GFX9 %s
+; RUN: llc -verify-machineinstrs -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1030 < %s | FileCheck -check-prefix=GFX10 %s
+; RUN: llc -verify-machineinstrs -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1100 < %s | FileCheck -check-prefix=GFX11 %s
+
+define protected amdgpu_kernel void @test(ptr addrspace(1) %srcA, ptr addrspace(1) %srcB, ptr addrspace(1) %dst) {
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/match-perm-extract-vector-elt-bug.ll:6]:
```suggestion
define amdgpu_kernel void @test(ptr addrspace(1) %srcA, ptr addrspace(1) %srcB, ptr addrspace(1) %dst) {
```

File: llvm/test/CodeGen/AMDGPU/match-perm-extract-vector-elt-bug.ll (hunk 2 above)
Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/match-perm-extract-vector-elt-bug.ll:106]: I reduced it further. This looks like the minimal version that can produce `v_perm_b32` instruction.

**Comment by shiltian** [comment]
@jrbyrnes I tried your case but didn't find anything wrong. I set a breakpoint at `SIISelLowering.cpp:12068`, which is `case ISD::EXTRACT_VECTOR_ELT`. The break point was hit twice.

Here is when it was hit the first time. It was trying to match `t64: i32 = or # D:1 t61, t63` and the index at this point is 3.

```
SelectionDAG has 45 nodes:
  t0: ch,glue = EntryToken
      t2: i32,ch = CopyFromReg # D:1 t0, Register:i32 %0
      t4: i32,ch = CopyFromReg # D:1 t0, Register:i32 %1
    t13: i64 = build_pair # D:1 t2, t4
  t18: v4i32,ch = load<(load (s128) from %ir.in0, align 4, addrspace 1)> # D:1 t0, t13, undef:i64
      t6: i32,ch = CopyFromReg # D:1 t0, Register:i32 %2
      t8: i32,ch = CopyFromReg # D:1 t0, Register:i32 %3
    t14: i64 = build_pair # D:1 t6, t8
  t19: v4i32,ch = load<(load (s128) from %ir.in1, align 4, addrspace 1)> # D:1 t0, t14, undef:i64
  t25: i32 = extract_vector_elt # D:1 t19, Constant:i32<3>
  t51: i16 = truncate # D:1 t25
      t29: ch = TokenFactor t18:1, t19:1
                  t22: i32 = extract_vector_elt # D:1 t18, Constant:i32<1>
                t36: i32 = srl # D:1 t22, Constant:i32<16>
              t37: i16 = truncate # D:1 t36
            t42: i16 = srl # D:1 t37, Constant:i16<8>
                t52: i32 = srl # D:1 t25, Constant:i32<16>
              t53: i16 = truncate # D:1 t52
            t89: i16 = and # D:1 t53, Constant:i16<-256>
          t74: i16 = or # D:1 t42, t89
        t61: i32 = zero_extend # D:1 t74
              t85: i16 = and # D:1 t51, Constant:i16<255>
              t82: i16 = shl # D:1 t51, Constant:i16<8>
            t83: i16 = or # D:1 t85, t82
          t62: i32 = any_extend # D:1 t83
        t63: i32 = shl # D:1 t62, Constant:i32<16>
      t64: i32 = or # D:1 t61, t63
        t10: i32,ch = CopyFromReg # D:1 t0, Register:i32 %4
        t12: i32,ch = CopyFromReg # D:1 t0, Register:i32 %5
      t15: i64 = build_pair # D:1 t10, t12
    t33: ch = store<(store (s32) into %ir.out0, addrspace 1)> # D:1 t29, t64, t15, undef:i64
  t31: ch = RET_GLUE t33
```

Here is the trace of the byte:

```
t64[3] -> t63[3] -> t62[1] -> t83[1] -> t82[1] -> t51[0] -> t25[0] -> t19[12]
```

This looks correct to me.

Here is when the break point was hit the second time. Still, it was trying to match `t64: i32 = or # D:1 t61, t63`, and the index is 3.

```
SelectionDAG has 45 nodes:
  t0: ch,glue = EntryToken
      t2: i32,ch = CopyFromReg # D:1 t0, Register:i32 %0
      t4: i32,ch = CopyFromReg # D:1 t0, Register:i32 %1
    t13: i64 = build_pair # D:1 t2, t4
  t18: v4i32,ch = load<(load (s128) from %ir.in0, align 4, addrspace 1)> # D:1 t0, t13, undef:i64
      t6: i32,ch = CopyFromReg # D:1 t0, Register:i32 %2
      t8: i32,ch = CopyFromReg # D:1 t0, Register:i32 %3
    t14: i64 = build_pair # D:1 t6, t8
  t19: v4i32,ch = load<(load (s128) from %ir.in1, align 4, addrspace 1)> # D:1 t0, t14, undef:i64
  t25: i32 = extract_vector_elt # D:1 t19, Constant:i32<3>
  t51: i16 = truncate # D:1 t25
      t29: ch = TokenFactor t18:1, t19:1
                t22: i32 = extract_vector_elt # D:1 t18, Constant:i32<1>
              t91: i32 = srl # D:1 t22, Constant:i32<24>
            t92: i16 = truncate # D:1 t91
                t52: i32 = srl # D:1 t25, Constant:i32<16>
              t53: i16 = truncate # D:1 t52
            t89: i16 = and # D:1 t53, Constant:i16<-256>
          t74: i16 = or # D:1 t92, t89
        t61: i32 = zero_extend # D:1 t74
              t85: i16 = and # D:1 t51, Constant:i16<255>
              t82: i16 = shl # D:1 t51, Constant:i16<8>
            t83: i16 = or # D:1 t85, t82
          t62: i32 = any_extend # D:1 t83
        t63: i32 = shl # D:1 t62, Constant:i32<16>
      t64: i32 = or # D:1 t61, t63
        t10: i32,ch = CopyFromReg # D:1 t0, Register:i32 %4
        t12: i32,ch = CopyFromReg # D:1 t0, Register:i32 %5
      t15: i64 = build_pair # D:1 t10, t12
    t33: ch = store<(store (s32) into %ir.out0, addrspace 1)> # D:1 t29, t64, t15, undef:i64
  t31: ch = RET_GLUE t33
```

The trace is still the same:

```
t64[3] -> t63[3] -> t62[1] -> t83[1] -> t82[1] -> t51[0] -> t25[0] -> t19[12]
```

**Comment by arsenm** [comment]
Should include the Fixes: SWDEV-... comment in description

**Comment by arsenm** [comment]
> I see a regression of class:
> 
> define hidden void @new(ptr addrspace(1) %in0, ptr addrspace(1) %in1, ptr addrspace(1) %out0) { %vec0 = load <4 x i32>, ptr addrspace(1) %in0, align 4 %vec1 = load <4 x i32>, ptr addrspace(1) %in1, align 4 %v0e1 = extractelement <4 x i32> %vec0, i64 1 %v1e3 = extractelement <4 x i32> %vec1, i64 3 %bc0 = bitcast i32 %v0e1 to <4 x i8> %bc1 = bitcast i32 %v1e3 to <4 x i8> %shuffle0_0 = shufflevector <4 x i8> %bc0, <4 x i8> %bc1, <4 x i32> <i32 3, i32 7, i32 4, i32 4> store <4 x i8> %shuffle0_0, ptr addrspace(1) %out0, align 4 ret void }

Is this one of the tests in the file? What do you mean by regression?

**Comment by shiltian** [comment]
> Should include the Fixes: SWDEV-... comment in description

Do we want to include that information in upstream PR?

**Comment by arsenm** [comment]
> > Should include the Fixes: SWDEV-... comment in description
> 
> Do we want to include that information in upstream PR?

Yes

**Comment by shiltian** [comment]
ping

**Comment by shiltian** [comment]
@jrbyrnes

**Comment by shiltian** [comment]
ping +1

**Comment by arsenm** [comment]
> > > Should include the Fixes: SWDEV-... comment in description
> > 
> > 
> > Do we want to include that information in upstream PR?
> 
> Yes

Specifically, the format "Fixes: SWDEV-xxxxxx"

**Comment by arsenm** [comment]
> Is this one of the tests in the file? What do you mean by regression?

What was this regression? Is this one of the existing tests?

**Comment by jrbyrnes** [comment]
> > Is this one of the tests in the file? What do you mean by regression?
> 
> What was this regression? Is this one of the existing tests?

Thanks for the ping -- this got buried in emails while I was away.

By regression, I mean quality of code, not correctness. Using the example provided

llc -mtriple=amdgcn -mcpu=gfx908 -verify-machineinstrs test.ll

git fetch origin pull/91697/head:fix-cbp
git checkout fix-cbp

(with PR)

        s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
        global_load_dword v6, v[2:3], off offset:12
        global_load_dword v7, v[0:1], off offset:4
        s_movk_i32 s4, 0xff00
        s_waitcnt vmcnt(1)
        v_lshlrev_b16_e32 v0, 8, v6
        v_and_b32_sdwa v1, v6, s4 dst_sel:DWORD dst_unused:UNUSED_PAD src0_sel:WORD_1 src1_sel:DWORD
        v_or_b32_sdwa v0, v6, v0 dst_sel:WORD_1 dst_unused:UNUSED_PAD src0_sel:BYTE_0 src1_sel:DWORD
        s_waitcnt vmcnt(0)
        v_or_b32_sdwa v1, v7, v1 dst_sel:DWORD dst_unused:UNUSED_PAD src0_sel:BYTE_3 src1_sel:DWORD
        v_or_b32_sdwa v0, v1, v0 dst_sel:DWORD dst_unused:UNUSED_PAD src0_sel:WORD_0 src1_sel:DWORD
        global_store_dword v[4:5], v0, off
        s_waitcnt vmcnt(0)
        s_setpc_b64 s[30:31]

git revert

(without PR)

        s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
        global_load_dword v6, v[0:1], off offset:4
        global_load_dword v7, v[2:3], off offset:12
        s_movk_i32 s4, 0x307
        s_waitcnt vmcnt(0)
        v_perm_b32 v0, v6, v7, s4
        global_store_dword v[4:5], v0, off
        s_waitcnt vmcnt(0)
        s_setpc_b64 s[30:31]

I have seen a similar pattern in existing real code (CK).  I don't think it is covered in existing lit tests.

SelectionDAG has 45 nodes:
  t0: ch,glue = EntryToken
      t2: i32,ch = CopyFromReg # D:1 t0, Register:i32 %0
      t4: i32,ch = CopyFromReg # D:1 t0, Register:i32 %1
    t13: i64 = build_pair # D:1 t2, t4
  t18: v4i32,ch = load<(load (s128) from %ir.in0, align 4, addrspace 1)> # D:1 t0, t13, undef:i64
      t6: i32,ch = CopyFromReg # D:1 t0, Register:i32 %2
      t8: i32,ch = CopyFromReg # D:1 t0, Register:i32 %3
    t14: i64 = build_pair # D:1 t6, t8
  t19: v4i32,ch = load<(load (s128) from %ir.in1, align 4, addrspace 1)> # D:1 t0, t14, undef:i64
  t25: i32 = extract_vector_elt # D:1 t19, Constant:i32<3>
  t51: i16 = truncate # D:1 t25
      t29: ch = TokenFactor t18:1, t19:1
                  t22: i32 = extract_vector_elt # D:1 t18, Constant:i32<1>
                t36: i32 = srl # D:1 t22, Constant:i32<16>
              t37: i16 = truncate # D:1 t36
            t42: i16 = srl # D:1 t37, Constant:i16<8>
                t52: i32 = srl # D:1 t25, Constant:i32<16>
              t53: i16 = truncate # D:1 t52
            t89: i16 = and # D:1 t53, Constant:i16<-256>
          t74: i16 = or # D:1 t42, t89
        t61: i32 = zero_extend # D:1 t74
              t85: i16 = and # D:1 t51, Constant:i16<255>
              t82: i16 = shl # D:1 t51, Constant:i16<8>
            t83: i16 = or # D:1 t85, t82
          t62: i32 = any_extend # D:1 t83
        t63: i32 = shl # D:1 t62, Constant:i32<16>
      t64: i32 = or # D:1 t61, t63
        t10: i32,ch = CopyFromReg # D:1 t0, Register:i32 %4
        t12: i32,ch = CopyFromReg # D:1 t0, Register:i32 %5
      t15: i64 = build_pair # D:1 t10, t12
    t33: ch = store<(store (s32) into %ir.out0, addrspace 1)> # D:1 t29, t64, t15, undef:i64
  t31: ch = RET_GLUE t33

With the PR, we have source bytes: t22, t25, t25, and t19. Without the PR, I see source bytes: t22, t22, t25, t25. The issue here seems to be too many source nodes, and trying to combine them is not profitable.

It seems the answer is that if we are going to look through extract_vector_elt in calculateByteProvider, we should also look them in calculateSrcByte. But, doing such may cause your crash to reappear (I don't understand how this is causing a crash?)

**Comment by shiltian** [comment]
It didn't cause a crash. It causes wrong results in just one OpenCL conformance test. That is because wrong register is used in `v_perm_b32` instruction.

Correct asm:

```
...
	v_perm_b32 v2, v2, v2, 0x3020504
	v_perm_b32 v0, v0, v0, 0x3020504
...
```

Wrong asm:

```
...
	v_perm_b32 v2, v0, v2, 0x3020706
	v_perm_b32 v0, v0, v0, 0x3020504
...
```

I narrowed it down to this place and found out that the index was wrong when calculating the source byte.

**Comment by shiltian** [comment]
> > It didn't cause a crash. It causes wrong results in just one OpenCL conformance test. That is because wrong register is used in `v_perm_b32` instruction.
> 
> Ah, I see the issue. Thanks.
> 
> It is because the current index calculation condition assumes less than 32 bit.
> 
> Can you try
> 
> ```
>     if (ScalarSize < 32) {
>       Index = ScalarSize == 8 ? VecIdx : VecIdx * 2 + Index;
>     }
> 
>     return calculateSrcByte(ScalarSize >= 32 ? Op : Op.getOperand(0),
>                             StartingIndex, Index);
> ```

It works. Why do we want to differentiate them here?

**Comment by jrbyrnes** [comment]
> Why do we want to differentiate them here?

We don't gain anything from looking into the vectors for scalarsize >= 32. calculateSrcByte will just return the Op since vector handling hasn't been built yet, and then we will need to re-extract the dword when building the perm https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AMDGPU/SIISelLowering.cpp#L12193  -- by disallowing them, we avoid some unecessary computation. Additionally, it does not introduce the regression mentioned above.

**Comment by shiltian** [comment]
> We don't gain anything from looking into the vectors for scalarsize >= 32. calculateSrcByte will just return the Op since vector handling hasn't been built yet, and then we will need to re-extract the dword when building the perm https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AMDGPU/SIISelLowering.cpp#L12193 -- by disallowing them, we avoid some unecessary computation. Additionally, it does not introduce the regression mentioned above.
> 
> On the other hand, we want to continue looking through to the vector for 8 and 16 bit scalars, as the vector that we are extracting from is usually the common source of permutation (and not the extracted value itself).

Good to know. Thanks for your explanation. I have updated the patch.

**Review by jrbyrnes — APPROVED** [review]
Thanks, LGTM

**Pull request merged**